From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Mon, 6 Apr 2020 12:16:56 +0200
Subject: [PATCH] PVE-Backup: modify job api

Introduce a pause_count parameter to start a backup in
paused mode. This way backups of multiple drives can be
started up sequentially via the completion callback while
having been started at the same point in time.

Signed-off-by: Thomas Lamprecht <t.lamprecht@proxmox.com>
---
 block/backup.c            | 3 +++
 block/replication.c       | 2 +-
 blockdev.c                | 3 ++-
 include/block/block_int.h | 1 +
 job.c                     | 2 +-
 5 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/block/backup.c b/block/backup.c
index a7a7dcaf4c..ecd93e91e0 100644
--- a/block/backup.c
+++ b/block/backup.c
@@ -338,6 +338,7 @@ BlockJob *backup_job_create(const char *job_id, BlockDriverState *bs,
                   BlockdevOnError on_target_error,
                   int creation_flags,
                   BlockCompletionFunc *cb, void *opaque,
+                  int pause_count,
                   JobTxn *txn, Error **errp)
 {
     int64_t len;
@@ -459,6 +460,8 @@ BlockJob *backup_job_create(const char *job_id, BlockDriverState *bs,
     block_job_add_bdrv(&job->common, "target", target, 0, BLK_PERM_ALL,
                        &error_abort);
 
+    job->common.job.pause_count += pause_count;
+
     return &job->common;
 
  error:
diff --git a/block/replication.c b/block/replication.c
index da013c2041..17246a822c 100644
--- a/block/replication.c
+++ b/block/replication.c
@@ -554,7 +554,7 @@ static void replication_start(ReplicationState *rs, ReplicationMode mode,
                                 0, MIRROR_SYNC_MODE_NONE, NULL, 0, false, NULL,
                                 BLOCKDEV_ON_ERROR_REPORT,
                                 BLOCKDEV_ON_ERROR_REPORT, JOB_INTERNAL,
-                                backup_job_completed, bs, NULL, &local_err);
+                                backup_job_completed, bs, 0, NULL, &local_err);
         if (local_err) {
             error_propagate(errp, local_err);
             backup_job_cleanup(bs);
diff --git a/blockdev.c b/blockdev.c
index 5faddaa705..65c358e4ef 100644
--- a/blockdev.c
+++ b/blockdev.c
@@ -3114,7 +3114,8 @@ static BlockJob *do_backup_common(BackupCommon *backup,
                             backup->filter_node_name,
                             backup->on_source_error,
                             backup->on_target_error,
-                            job_flags, NULL, NULL, txn, errp);
+                            job_flags, NULL, NULL, 0, txn, errp);
+
     return job;
 }
 
diff --git a/include/block/block_int.h b/include/block/block_int.h
index 4c3587ea19..336f71e69d 100644
--- a/include/block/block_int.h
+++ b/include/block/block_int.h
@@ -1219,6 +1219,7 @@ BlockJob *backup_job_create(const char *job_id, BlockDriverState *bs,
                             BlockdevOnError on_target_error,
                             int creation_flags,
                             BlockCompletionFunc *cb, void *opaque,
+                            int pause_count,
                             JobTxn *txn, Error **errp);
 
 BdrvChild *bdrv_root_attach_child(BlockDriverState *child_bs,
diff --git a/job.c b/job.c
index 53be57a3a0..e82253e041 100644
--- a/job.c
+++ b/job.c
@@ -918,7 +918,7 @@ void job_start(Job *job)
     job->co = qemu_coroutine_create(job_co_entry, job);
     job->pause_count--;
     job->busy = true;
-    job->paused = false;
+    job->paused = job->pause_count > 0;
     job_state_transition(job, JOB_STATUS_RUNNING);
     aio_co_enter(job->aio_context, job->co);
 }
